src/hunspell/csutil.cxx

```diff
@@ -2418,6 +2420,7 @@ int get_lang_num(const std::string& lang) {
 #ifndef OPENOFFICEORG
 #ifndef MOZILLA_CLIENT
 void initialize_utf_tbl() {
+  std::lock_guard<std::mutex> guard(utf_tbl_mutex);
```

Comment:
I think it would be possible to use an atomic counter to avoid using a full mutex in common case that it is already initialized. Possibly something like:

```c++
if (utf_tbl_count.load(memory_order_acquire) == 0) {
    utf_tbl_count.fetch_add(1, memory_order_release);
    return;
}
std::lock_guard<std::mutex> guard(utf_tbl_mutex);
utf_tbl = new unicode_info2[CONTSIZE];
// etc.
utf_tbl_count.fetch_add(1, memory_order_release); // do this at the end so that other threads will also have to wait on the lock for the table to be initialized.
}
```
and change `free_utf_tbl` to:

```c++
if (utf_tbl_count.load(memory_order_acquire) == 1) {
    std::lock_guard<std::mutex> guard(utf_tbl_mutex);
    delete[] utf_tbl;
    utf_tbl = NULL;
    utf_tbl_count.store(0, memory_order_release);
} else {
    utf_tbl_count.fetch_sub(1, memory_order_release);
}
```

Or possibly the utf table could be stored in thread local storage? although then that would require more memory and computation to create the tables.